### PR TITLE
ruby: Replace default tasks with a stub message

### LINKED ISSUE #12579: Ruby extension incorrectly detects tests
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

The queries used to detect Minitest tests, RSpec example groups and RSpec specs are not sufficient to be included by default in the Ruby extension. Simply calling the method `test`, `it` or `describe` does not guarantee you are using Minitest or RSpec. Additionally, the definition of a method starting with `test_` does not guarantee the use of Minitest.

There are other test libraries for Ruby including [Sus](https://github.com/ioquatix/sus), [TLDR](https://github.com/tendersearls/tldr), and [Quickdraw](https://github.com/joeldrapper/quickdraw) that conflict with these queries.

Additionally, the commands that the runnables tasks trigger are not even correct for RSpec or Minitest. For RSpec, the command should start with `bundle exec rspec` to ensure the bundled version of RSpec is used. For minitest, this is more complicated. It’s usually `bundle exec rake` or `bundle exec rake test`. Sometimes it’s `bundle exec rails test` and sometimes (when working on a Rails gem), it’s `bin/test`.

## Steps to Reproduce
In any Ruby file, these three examples trigger runnables without correct scoping:

```ruby
def test_something
end

describe Something

test "something"
```

## Recommended solution

At minimum, the runnables should be removed by default. Perhaps they could be enabled with a configuration or by installing an `rspec` or a `minitest` extension. Alternatively, it might be possible to write more precise queries to detect the presence of the relevant Ruby gems.

I don’t know if this is possible, but the *ideal* solution would probably be to allow project-level configuration of tasks with the relevant runnable detection queries. That way, users could copy/paste e.g. the Minitest configuration into their project and update the task to execute `bundle exec rake` or `bin/test` or whatever they need.

### Environment

Zed: v0.137.6 (Zed)
OS: macOS 14.4.1
Memory: 128 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by vitallium**
The Ruby world has many testing frameworks:

- Minitest
- RSpec
- quickdraw
- tldr
- and many others.

Attempting to support all of them through a single `tasks.json` file is a challenging task and nearly impossible. All testing frameworks have different running options and commands. It's still possible to use tree-sitter queries to detect runnables in Ruby code but Zed lacks the ability to detect the testing framework in a project that can be used to detect the correct commands to run tests or runnables. The end user knows the correct command and it's wise to delegate creating the command to them. It would be a bit strange to leave the user without any guidance, so this commit adds example tasks for various Ruby testing frameworks.

Closes #12579

Here is the screenshot how it looks:

CleanShot 2024-07-01 at 19 37 08@2x

Release Notes:

- N/A

### CODE REVIEW

**Comment by osiewicz** [comment]
Hey,
Today I've landed support for user-defined task variables in https://github.com/zed-industries/zed/pull/13699, which should pave way for having tasks tailored for a particular testing framework. While it doesn't support languages provided by extensions, would ability to tailor returned tasks help you anyhow?
I see it as a bit of a failure of tasks if we can't make your workflow work, which I'm sorry for.

**Comment by vitallium** [comment]
> Hey, Today I've landed support for user-defined task variables in #13699, which should pave way for having tasks tailored for a particular testing framework. While it doesn't support languages provided by extensions yet, would ability to tailor returned tasks help you anyhow? I see it as a bit of a failure of tasks if we can't make your workflow work, which I'm sorry for.

@osiewicz, oh, that looks interesting! I will check how that works and get back to you. Thanks for the heads-up!

**Comment by vitallium** [comment]
>While it doesn't support languages provided by extensions yet, would ability to tailor returned tasks help you anyhow?

@osiewicz I don't have a strong opinion on that. I think, since the linked pull request (https://github.com/zed-industries/zed/pull/13699) states that user-defined task variables are an experimental feature, this feature can be changed at any point by introducing breaking changes in the extensions API. Additionally, Ruby is not limited by testing frameworks but also by the commands that we use to run them: `bundle exec`, `./bin/rake`, just `rspec` or something else. The proposed workaround for Ruby with user-defined tasks seems more or less fine to provide a flexible solution for all problems. All in all, I am pretty confident that you, Zed engineers, will provide the best solution for handling runnables and user tasks at some point. And then the Ruby extension will utilize this API to handle all cases. Plus, I know that you are extremely busy with other features, and this feature does not look like a must-have for Ruby. I feel it would be better to let you focus on other things. Thanks!

**Comment by osiewicz** [comment]
That's fair, though just to remark, in marking it as experimental I wanted to highlight how it shouldn't be mentioned in release notes just yet. We'd have to make a breaking change to extension API anyways in order to introduce it for extensions.

**Comment by vitallium** [comment]
> That's fair, though just to remark, in marking it as experimental I wanted to highlight how it shouldn't be mentioned in release notes just yet. We'd have to make a breaking change to extension API anyways in order to introduce it for extensions.

Ah, in this case, if you are still willing to help, we can explore that path. Just let me know if you need anything. Thank you for your help!

**Comment by vitallium** [comment]
@joeldrapper Hi, would it be possible for you to check this PR and provide any tips and guidance on improving the documentation for running Ruby tests? Thanks!

**Comment by joeldrapper** [comment]
👍 I’ll take a look today.

**Comment by SomeoneToIgnore** [comment]
This looks quite stale, so I'll close the PR to have less items in the queue.

File: docs/src/languages/ruby.md [hunk 1]
```diff
@@ -126,3 +126,59 @@ end
...
+]
+```
+
+### quickdraw
+
+```json
+[
+  {
+    "label": "test $ZED_RELATIVE_FILE:$ZED_ROW",
+    "command": "./bin/qt",
```
Comment (joeldrapper) [inline docs/src/languages/ruby.md:166]: I don't think this is right unless there’s some separate process where binstubs are generated. Quickdraw is usually run with `bundle exec qt {file}`.

File: docs/src/languages/ruby.md [hunk 2]
```diff
@@ -126,3 +126,59 @@ end
...
+
+To run tests in your Ruby project, you can set up custom tasks in your local `.zed/tasks.json` configuration file. These tasks can be defined to work with different test frameworks like Minitest, RSpec, quickdraw, and tldr. Below are some examples of how to set up these tasks to run your tests from within your editor.
+
+### Minitest
+
+```json
+[
+  {
+    "label": "test $ZED_RELATIVE_FILE:$ZED_ROW",
+    "command": "./bin/rails",
```
Comment (joeldrapper) [inline docs/src/languages/ruby.md:140]: I think `bin/rails` probably does work but it’s non-standard. I would expect this to be `bundle exec rails test {file}`.

**Comment by joeldrapper** [comment]
Sorry, I completely forgot about this PR until @SomeoneToIgnore reminded me. Left a couple of comments.

File: docs/src/languages/ruby.md (hunk 1 above)
Comment (vitallium) [reply docs/src/languages/ruby.md:166]: @joeldrapper I used a `binstub` here because the `command` requires an absolute path. But I am not sure if this is still a thing. I will double-check that.

Comment (vitallium) [reply docs/src/languages/ruby.md:166]: Heeey, it actually works!

CleanShot 2024-08-23 at 16 13 08@2x


I will update example in this PR to `bundle exec` then. Thank you, @joeldrapper!

**Comment by vitallium** [comment]
> Sorry, I completely forgot about this PR until @SomeoneToIgnore reminded me. Left a couple of comments.

@joeldrapper no worries! I re-opened it [here](https://github.com/zed-industries/zed/pull/16752) with your suggestions.
